tests/http/server/_base.py

```diff
@@ -98,3 +111,413 @@ def _test_disconnect(
+                                self.fail(
+                                    "Request appears to be stuck, possibly due to "
+                                    "a previous cancelled request"
+                                )
+
+                    if respond_mock.called:
+                        # The request ran to completion and we are done with testing it.
+
+                        # `respond_with_json` writes the response asynchronously, so we
+                        # might have to give the reactor a kick before the channel gets
+                        # the response.
+                        reactor.pump([1.0])
+
+                        self.assertEqual(channel.code, expected_code)
+                        return channel.json_body
+                    else:
+                        # Disconnect the client and wait for the response.
+                        request.connectionLost(reason=ConnectionDone())
+
+                        self._log_for_request(request_number, "--- disconnected ---")
+
+                        # We may need to pump the reactor to allow `delay_cancellation`s
+                        # to finish.
+                        if not respond_mock.called:
+                            reactor.pump([0.0])
+
+                        # Try advancing the clock if that didn't work.
+                        if not respond_mock.called:
+                            reactor.pump([1.0])
+
+                        # Mark the request's logging context as finished. If it gets
+                        # activated again, an `AssertionError` will be raised. This
+                        # `AssertionError` will likely be caught by twisted and turned
+                        # into a `Failure`. Instead, a different `AssertionError` will
+                        # be observed when the logging context is deactivated, as it
+                        # wouldn't have tracked resource usage correctly.
+                        if isinstance(request, SynapseRequest) and request.logcontext:
+                            request.logcontext.finished = True
+
+                        # Check that the request finished with a 499,
+                        # ie. the `CancelledError` wasn't swallowed.
+                        respond_mock.assert_called_once()
+                        args, _kwargs = respond_mock.call_args
+                        code = args[1]
+                        self.assertEqual(code, HTTP_STATUS_REQUEST_CANCELLED)
+            finally:
+                # Unblock any processing that might be shared between requests.
+                unblock_awaits()
+
+        assert False, "unreachable"  # noqa: B011
+
+    def create_deferred___next___patch(
+        self, seen_awaits: Set[Tuple[str, ...]], request_number: int
+    ) -> Tuple[
+        Callable[["Deferred[T]"], "Deferred[T]"],
+        Callable[[], None],
+        Callable[[], int],
+        Callable[[], bool],
+    ]:
```

Comment:
Suggest making this whole thing an object, and turn `create_deferred___next___patch` into its constructor.